File: go/splitwise/split.go
package splitwise

// SimplifyDebts takes a slice of transactions between parties and
// returns the minimum number of payments between parties necessary
// to settle the outstanding debts
func SimplifyDebts(transactions Transactions) (simplifiedDebts Transactions) {

	// compute total balances
	scores := Scores(transactions)

	if !isZeroSum(scores) {
		panic("invalid scores, must be zero sum")
	}

	return greedy(scores, simplifiedDebts)
}

func isZeroSum(scores map[int64]int64) bool {
	var v int64
	for i := range scores {
		v += scores[i]
	}
	return int(v) == 0
}

// greedy is a recursive algorithm that simplifies debts
// within the group given the current set of outstanding balances.
// The algorithm is as follows
//
// Step 1)
//
//	Work out maximum creditor and debtor minimum and maximum scores
//
//		    mD - minimum score, the largest debtor in the group
//		    mC - maximum score, the largest creditor in the group
//
//	        maxDebtor - index for debtor with score mD
//	        maxCreditor - index for creditor with score mC
//
// Step 2)
//
// if |mD| == |mC| then create a transaction from maxDebtor to maxCreditor of mC and eliminate both maxDebtor and maxCreditor from the group
// if |mD| <  |mC| then create a transaction from maxDebtor to maxCreditor of |mD| and eliminate maxDebtor from the group
// if |mD| >  |mC| then create a transaction from maxDebtor to maxCreditor of mC and eliminate maxCreditor from the group
//
// Step 3)
//
//	Repeat Step 1) - 2) until the group is eliminated and return
//
// WARNING: Does not protect against malicious inputs,
// To avoid an infinite loop use isZeroSum() to check if the input scores are valid
// before executing this function
func greedy(scores map[int64]int64, existingDebts Transactions) (debts Transactions) {

	if len(scores) == 0 {
		return existingDebts
	}

	maxCreditor, mC := maxEntry(scores)

	maxDebtor, mD := minEntry(scores)

	if mC == -mD {
		debts = Add(existingDebts, maxDebtor, maxCreditor, mC)
		delete(scores, maxDebtor)
		delete(scores, maxCreditor)
	} else if mC > -mD {
		debts = Add(existingDebts, maxDebtor, maxCreditor, -mD)
		delete(scores, maxDebtor)
		scores[maxCreditor] += mD
	} else if mC < -mD {
		debts = Add(existingDebts, maxDebtor, maxCreditor, mC)
		delete(scores, maxCreditor)
		scores[maxDebtor] += mC
	}

	return greedy(scores, debts)
}

func maxEntry(m map[int64]int64) (index, value int64) {
	// find first element
	for i := range m {
		index = i
		value = m[index]
		break
	}
	// obtain max element
	for x := range m {
		if m[x] > value {
			index, value = x, m[x]
		}
	}
	return
}

func minEntry(m map[int64]int64) (index, value int64) {
	// find first element
	for i := range m {
		index = i
		value = m[index]
		break
	}
	// obtain min element
	for x := range m {
		if m[x] < value {
			index, value = x, m[x]
		}
	}
	return
}

File: python/src/splitwise/split.py
from collections import defaultdict
from array import *
import numpy as np
import time
"""
This programme takes an input file containing transactions between nodes in a graph,
calculates the total balances, simplifies the debts and prints out the minimum transactions
required to settle those debts.

For examples of valid input csv files see ../../../test_data

Requirements: Python 3.6 (or later), numpy

Install NumPy using pip --> $ python3 -m pip install --user numpy 
"""


def greedy(scores, debt_array):
    if len(scores) == 0:
        return debt_array
        
    max_creditor, max_credit = max_entry(scores)
    max_debtor, max_debt = min_entry(scores)

    new_debt_array = debt_array.copy()

    if max_credit == -max_debt:
        new_debt_array.append([max_debtor, max_creditor, -max_debt])
        del scores[max_debtor]
        del scores[max_creditor]
    elif max_credit  > -max_debt:
        new_debt_array.append([max_debtor, max_creditor, -max_debt])
        del scores[max_debtor]
        scores[max_creditor] += max_debt
    elif max_credit  < -max_debt:
        new_debt_array.append([max_debtor, max_creditor, max_credit])
        del scores[max_creditor]
        scores[max_debtor] += max_credit

    updated_scores = scores
    return greedy(updated_scores, new_debt_array)

def scores(transactions):
    scores = defaultdict(int)

    for f, t, a in transactions:
        scores[f] -= a
        scores[t] += a

    return scores


def max_entry(m):
	# find first element
	for i in m: 
		index = i
		value = m[index]
		break
	
	# obtain max element
	for x in m:
		if m[x] > value:
			index, value = x, m[x]
	return index, value


def min_entry(m):
	# find first element
	for i in m :
		index = i
		value = m[index]
		break
	# obtain min element
	for x in m :
		if m[x] < value :
			index, value = x, m[x]
	return index, value

def is_zero_sum(scores):
    v = 0
    for i in scores:
        v += scores[i]
    return v == 0

"""
Solution

Class containing minimum transaction algorithm and csv decoding 
"""
class Solution: 

    def __init__(self):      
        # do nothing
        return

    def read_file(self, input_csv):
        return np.genfromtxt(input_csv, delimiter=',')

    def simplify_debts(self, transactions):
        balances = scores(transactions)
        if is_zero_sum(balances) != True:
            raise Exception("invalid scores, must be zero sum")
        debts = list()
        return greedy(balances, debts)




if __name__ == "__main__":
    print("splitwise algorithm in python")

    # instantiate class instance
    solution_instance = Solution

    # Measure csv load time
    start = time.perf_counter_ns()
    data = solution_instance().read_file("../../../test_data/input.csv")
    print(f"read csv file in {(time.perf_counter_ns() - start)/1000} microseconds")

    # Measure algorithm time
    restart = time.perf_counter_ns()
    m = solution_instance().simplify_debts(data)
    print(f"completed simplify_debts execution in {(time.perf_counter_ns() -  restart)/1000} microseconds, transactions {len(m)}")

    print("simplified debts", m)